### PR TITLE
rough filter UI

### PR DESCRIPTION
**Pull request by aidankmcl**
## Reviewers: 

- [ ] @sihrc 
## What's the purpose of the PR?

(I think card got archived)
This is to add filtering to the demo in terms of both source of data and the time period being collected. It seems that a bunch aren't being filtered by date on elastic search's side so will need to check that out. Also allows for tweaking of filtering and DBScan settings so that the user can tweak to suit their query.
## Description of changes

(Meta note: the what code was changed part of this might be a little redundant with all the diffs below)
A quick explanation of the stuff unrelated to filtering:
`__init__.py` messing with the the keywords stuff for better titles,
`clusterer.py` tweaking granularity of steps in DBScan epsilon,
Most everything else is related to filtering, might be some small UI tweaks, CSS stuff.
## Keep in mind while reviewing code:
- Is relevant code tested?
- Are added functions/methods documented?
- Separation of concerns (SOC)
- Don't repeat yourself (DRY) 
- Limit the number of positional arguments
- Is function/method length reasonable?
- Can code be broken down into smaller components?
- Where do added functions/methods belong?
- Are variable names descriptive?
- Are errors handled appropriately?
- Can logic be simplified?
- Does the code make sense in context? (expand the diff)

### CODE REVIEW

File: cluster/search/client.py [hunk 1]
```diff
@@ -114,7 +114,8 @@ def search(self, query, start_date=None, end_date=None, limit=100, only_document
             u'timed_out': False
         }
         """
-        results = self.es.search(index=self.index, track_scores=True, body={
+        print source
```
Comment (sihrc) [inline cluster/search/client.py]: print statement

File: cluster/search/client.py [hunk 2]
```diff
@@ -124,17 +125,30 @@ def search(self, query, start_date=None, end_date=None, limit=100, only_document
...
+            "highlight" : {
+                "pre_tags" : ["<em>"],
+                "post_tags" : ["</em>"],
+                "fields" : {
+                    "_all" : {}
                 }
             }
-        }, size=limit, **kwargs)
+        }
+        if source:
```
Comment (sihrc) [inline cluster/search/client.py]: should be a part of the body. "*" wildcharacter matches all.

File: cluster/search/client.py [hunk 3]
```diff
@@ -124,17 +125,30 @@ def search(self, query, start_date=None, end_date=None, limit=100, only_document
...
+                                    "gte": start_date or EPOCH,
+                                    "lt": end_date or datetime.datetime.now()
+                                }
                             }
                         }
-                    }
+                    ]
+                },
+            },
+            "highlight" : {
```
Comment (sihrc) [inline cluster/search/client.py]: what is this doing? shouldn't fields=_all? or is the default for _all in the dict to highlight everything?

File: cluster/clustering/clusterer.py [hunk 4]
```diff
@@ -12,10 +12,11 @@ def __init__(self, feature_vectors, algorithm="brute", metric="cosine", **kwargs
             raise ClusterError('empty results')
         kwargs.update({
             "algorithm": algorithm,
-            "metric": "cosine"
+            "metric": metric
         })
 
-    def get_clusters(self, eps_range=[0.001, 0.002, 0.003, 0.004, 0.005, 0.006, 0.007, 0.008, 0.009, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, .1]):
+    def get_clusters(self, eps_range=[0.001, 0.002, 0.003, 0.0035, 0.004, 0.0045, 0.005, 0.0055, 0.006, 0.007, 0.008, 0.009, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, .1]):
+    # def get_clusters(self, eps_range=[0.05, 0.06, 0.07, 0.08, 0.09, 0.1, 0.11, 0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.2]):
```
Comment (sihrc) [inline cluster/clustering/clusterer.py]: commented

File: cluster/clustering/clusterer.py [hunk 5]
```diff
@@ -12,10 +12,11 @@ def __init__(self, feature_vectors, algorithm="brute", metric="cosine", **kwargs
             raise ClusterError('empty results')
         kwargs.update({
             "algorithm": algorithm,
-            "metric": "cosine"
+            "metric": metric
         })
 
-    def get_clusters(self, eps_range=[0.001, 0.002, 0.003, 0.004, 0.005, 0.006, 0.007, 0.008, 0.009, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, .1]):
+    def get_clusters(self, eps_range=[0.001, 0.002, 0.003, 0.0035, 0.004, 0.0045, 0.005, 0.0055, 0.006, 0.007, 0.008, 0.009, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, .1]):
```
Comment (sihrc) [inline cluster/clustering/clusterer.py:18]: perhaps, we should just fix this to be a required input.

File: cluster/clustering/__init__.py [hunk 6]
```diff
@@ -64,15 +65,19 @@ def _filter_threshold_keywords(result_dict):
         keywords_list.update(filter(lambda x: x != "Shutterstock", cluster_info['keywords']))
 
     for cluster, cluster_info in result_dict.items():
-        cluster_info['keywords'] = [value for value in cluster_info['keywords']
-                              if keywords_list[value] <= max(len(result_dict)*.35, 1)][-3:]
-        cluster_info['title_keywords'] = [value for value in cluster_info['title_keywords']
-                              if title_keywords_list[value] <= max(len(result_dict)*.35, 1)][-3:]
+        # cluster_info['keywords'] = [value for value in cluster_info['keywords']
```
Comment (sihrc) [inline cluster/clustering/__init__.py]: commented

File: cluster/clustering/__init__.py [hunk 7]
```diff
@@ -64,15 +65,19 @@ def _filter_threshold_keywords(result_dict):
...
+        cluster_info['keywords'] = cluster_info['keywords'][:10]
+        cluster_info['title_keywords'] = cluster_info['title_keywords'][:10]
         result_dict[cluster] = cluster_info
     return result_dict
 
 def _create_full_cluster_list(cluster_info, key):
-    return [entity for article in cluster_info['articles'] for entity in article['indico'].get(key)]
+    ents = [entity for article in cluster_info['articles']
+        for entity in article['indico'].get(key)]
+    return [e for e in ents if str(e['text']) not in BAD_ENTITIES]
```
Comment (sihrc) [inline cluster/clustering/__init__.py]: why isn't this if statement just part of the first comprehension?

File: cluster/app.py [hunk 8]
```diff
@@ -46,24 +47,35 @@ class QueryHandler(tornado.web.RequestHandler):
...
+            min_samples = int(data.get('min-samples', 3))
+            print min_samples, article_limit, ES_score_threshold, end, start, source
+
             self.set_secure_cookie('current_search', query)
 
-            entries = ES.search(query, start_date=before, end_date=after, limit=1000)
+            entries = ES.search(query, source=source, start_date=start, end_date=end, limit=article_limit)
+            print len(entries), entries[0]['score']
+            entries = filter(lambda x: x.get('score') > ES_score_threshold, entries)
+            print len(entries)
```
Comment (sihrc) [inline cluster/app.py]: print

File: cluster/app.py [hunk 9]
```diff
@@ -46,24 +47,35 @@ class QueryHandler(tornado.web.RequestHandler):
...
+            ES_score_threshold = float(data.get('threshold', 0.3))
+            article_limit = int(data.get('limit', 500))
+            min_samples = int(data.get('min-samples', 3))
+            print min_samples, article_limit, ES_score_threshold, end, start, source
+
             self.set_secure_cookie('current_search', query)
 
-            entries = ES.search(query, start_date=before, end_date=after, limit=1000)
+            entries = ES.search(query, source=source, start_date=start, end_date=end, limit=article_limit)
+            print len(entries), entries[0]['score']
```
Comment (sihrc) [inline cluster/app.py]: print

File: cluster/app.py [hunk 10]
```diff
@@ -46,24 +47,35 @@ class QueryHandler(tornado.web.RequestHandler):
...
-            before = data.get('before')
-            after = data.get('after')
+
+            source = data.get('source')
+            start = datetime(int(data.get('start', 2000)), 1, 1)
+            end = datetime(int(data.get('end', 2016)), 12, 31)
+            ES_score_threshold = float(data.get('threshold', 0.3))
+            article_limit = int(data.get('limit', 500))
+            min_samples = int(data.get('min-samples', 3))
+            print min_samples, article_limit, ES_score_threshold, end, start, source
```
Comment (sihrc) [inline cluster/app.py]: print

File: cluster/app.py [hunk 11]
```diff
@@ -46,24 +47,35 @@ class QueryHandler(tornado.web.RequestHandler):
     """
     def get(self):
         # Needs to be reworked to grab possible sources
-        self.write(json.dumps({'meow': 'meow'}))
+        sources = ES.get_uniques('source')
```
Comment (sihrc) [inline cluster/app.py]: try catch for possible errors.

File: cluster/search/client.py [hunk 12]
```diff
@@ -81,7 +81,7 @@ def count(self, query, **options):
         })
         return self.es.count(index=self.index, body=options)["count"]
 
-    def search(self, query, start_date=None, end_date=None, limit=100, only_documents=True, **kwargs):
+    def search(self, query, source=None, start_date=None, end_date=None, limit=100, only_documents=True, **kwargs):
```
Comment (sihrc) [inline cluster/search/client.py]: probably shoudl just set source default to be "*"

**Comment by sihrc** [comment]
We should add a preliminary integration test that tests a query we know the outcome for with some kind of heuristic. It would make checking the results of this much more efficient.

**Comment by sihrc** [comment]
We need tests for the changes to the client.py. There are existing tests.

File: cluster/search/client.py (hunk 3 above)
Comment (aidankmcl) [inline cluster/search/client.py]: This actually got left in by accident when testing to see if it did what I thought it did

File: cluster/search/client.py (hunk 12 above)
Comment (aidankmcl) [inline cluster/search/client.py]: Wildcard doesn't seem to work for me, did the part where it was getting added look correct? Not sure what's up with that

File: cluster/clustering/__init__.py [hunk 13]
```diff
@@ -7,23 +7,24 @@
...
             continue
         count[all_clusters[i]] += 1
-        if count[all_clusters[i]] > 20:
-            continue
         entry, cluster = entries[i], all_clusters[i]
         entry['cluster'], entry["distance"] = cluster, all_similarities[i]
         cluster_features[cluster] = cluster_features.get(cluster, []) + [feature_vectors[i]]
         result_dict[cluster]["articles"] = result_dict[cluster].get("articles", []) + [entry]
 
+    [result_dict.pop(cluster_num) for cluster_num, cluster in result_dict.items() if len(cluster.get('articles')) < 2]
```
Comment (sihrc) [inline cluster/clustering/__init__.py:26]: Hm, maybe something like
`result_dict = {cluster_num: cluster for cluster_num, cluster in result_dict.iteritems() if len(cluster.get("articles")) >= 2}`
might be a bit more intuitive to read.

**Pull request merged**